### ISSUE TITLE
Create a spacecraft details page

### ISSUE THREAD

**Issue by darozak** [Darwin-River/Ex-Machinis#78]
We also need to create a page which displays details for specific spacecraft.  This page will include tables that show the current contents of the done's cargo bay and list the most recent events that have affected the ship.  Here is an example of what the page should look like:
![200203 Spacecraft Details Page](https://user-images.githubusercontent.com/33500329/73655411-01d6a080-465c-11ea-96c0-1e39321327c0.PNG)

The Cargo Manifest was created with the following query:
```
SELECT agents.agent_id AS DroneID, First(agents.name) AS Drone, resources.name AS Resource, events.locked, First(events.new_quantity) AS Units, First(resources.mass) AS UnitMass, First([new_quantity]*[mass]) AS TotalMass
FROM ((agents LEFT JOIN events ON agents.agent_id = events.drone) LEFT JOIN resources ON events.resource = resources.id) LEFT JOIN event_types ON events.event_type = event_types.id
GROUP BY agents.agent_id, resources.name, events.locked
HAVING (((First(events.event_type))=1 Or (First(events.event_type))=2))
ORDER BY agents.agent_id, First(events.timestamp) DESC;
```

The events table uses the following query, which is essentially the same as the query we used for the events page.  In fact it could probably be simplified quite a bit because not all the data is displayed on the spacecraft details page:
```
SELECT observations.drone AS Observer, observations.Time AS ObservationTime, events.timestamp AS [Time], agents_1.user_id AS ActingCompanyID, users_1.name AS ActingCompanyName, agents_1.agent_id AS ActingDroneID, agents_1.name AS ActingDroneName, protocols.name AS [Action], event_types.name AS Effect, agents.user_id AS AffectedCompanyID, users.name AS AffectedCompanyName, agents.agent_id AS AffectedDroneID, agents.name AS AffectedDroneName, events.resource AS AffectedResourceID, resources.name AS AffectedResourceName, events.locked, events.new_quantity AS NewQuantity, events.new_credits, events.new_location AS NewLocationID, objects.object_name AS NewLocationName, Now() AS Now
FROM observations LEFT JOIN (((((((((events LEFT JOIN event_types ON events.event_type = event_types.id) LEFT JOIN objects ON events.new_location = objects.object_id) LEFT JOIN resources ON events.resource = resources.id) LEFT JOIN actions ON events.action = actions.id) LEFT JOIN protocols ON actions.protocol = protocols.id) LEFT JOIN agents AS agents_1 ON actions.drone = agents_1.agent_id) LEFT JOIN agents ON events.drone = agents.agent_id) LEFT JOIN users ON agents.user_id = users.user_id) LEFT JOIN users AS users_1 ON agents_1.user_id = users_1.user_id) ON observations.event = events.id
WHERE (((observations.drone)=0) AND ((observations.Time)<=Now()+1))
ORDER BY events.timestamp DESC;
```

**Comment by freelancefrivas** [comment]
So these are events that were observed by the ship? Is that what you wanted to represent with `(((observations.drone)=0)`? Or what column links to the drone?

**Comment by freelancefrivas** [comment]
You can take a look now at the spacecraft page (and I should take a second look tomorrow).
Some notes:
- I tried to put the cargo manifest and spacecraft details all in one row as in your mockup... but I didn't think it looked good. Although I may try again.
- You'll see I tried to condense columns. If you like the way the 'effect' includes the credits, location or units (when one or more of those fields are different than null), I can implement it in the main events table too...
- Ah, now that I write this I realize the cargo shows repeated resources in different items... I don't think that's right. I'll review it tomorrow.

**Comment by darozak** [comment]
The cargo manifest and events tables each involve fairly complex queries so it may take a bit of work to get these right.  Here are my thoughts on what you have done so far.

With respect to the cargo manifest:

- When generating the cargo manifest you should be grouping the results by agents.agent_id, resources.name, and events.locked and displaying only the first new_quantity value of each group.  Otherwise, you'll see multiple entries for each of the different resources.

- I'd like to break units and mass into separate columns for this table.

With respect to the events table:

- You're correct that  we should require the **observation.drone** to equal the **drone ID** and not zero. Thanks for catching this mistake.

- In addition to the acting spacecraft, we should add a column for the affected spacecraft.  Sorry that I forgot to add this earlier.

- Events should be displayed newest to oldest.

- I like the new way that you're displaying the events by combining the effects column into one! I changed the effects descriptions to support this way of displaying the data.  For example, I changed "Increment Inventory" to "Increment inventory to".  Can you remove the parens that are being placed around the quantities and locations?

- Can you make the spacecraft events table searchable just like you did for the main events table?

Once you've made these changes we can take another look and see how things are working.

**Comment by freelancefrivas** [comment]
I'm assuming you want it to be searchable across all the ship's events and not just the latest?

**Comment by darozak** [comment]
Yes.  That's correct.

**Comment by freelancefrivas** [comment]
What is observations.drone = 0 supposed to mean? Is this still required in the main game events table?

**Comment by freelancefrivas** [comment]
In the end I used the same JS component I had used in the events page... which is maybe what I should have done in the first place. You should check it...

**Comment by darozak** [comment]
You should set obsertations.drone equal to the spacecraft ID of the spacecraft whose page you're viewing.  When obsertations.drone=0 it means you're seeing events from a global perspective rather than that of a particular spacecraft.

Also, you don't need to display the company names in this events table because players won't be attempting to filter by company name.

Otherwise, it looks good.

**Comment by freelancefrivas** [comment]
Ok, I was already setting observations.drone = 0 when no agent is set, so that's good.
I removed the companies for this version of the table. I think I should also remove the sorting, as it doens't get along well with displaying only the last 10 events...

**Comment by darozak** [comment]
I just want to confirm that when I go to https://advolition.com/spacecraft/2 (Spacecraft ID =2), the events query sets observations.drone=2 and not zero.

Also, you said that you removed the company names but I'm still seeing them despite refreshing my browser.

I agree that we don't need the sorting.

Could you make the events table filter automatically as you type the keyword?  I like the way you did this for the [protocols page](https://advolition.com/protocols).

**Comment by freelancefrivas** [comment]
Sorry, I forgot to upload the latest app.js and app.css.
Ok, I removed the sorting.
The thing is, the protocols table is different. I'm not using the [vuetable-2 component](https://github.com/ratiw/vuetable-2) there. I'm filtering within the listed rows. In the events tables and most of the others, I used a button to submit the search in order to avoid sending a new request to the server every time the visitor presses a new key.
So ok, I can do it, but there is a cost: there will be more traffic, as I have to send a request for every key pressed...
